File: apps/server/src/api/profile-chat.ts
// Module: Exposes the authenticated profile collection chat endpoint.
import { Elysia } from "elysia"

import { ErrorResponseSchema } from "../schemas/error"
import {
  ProfileChatRequestSchema,
  ProfileChatResponseSchema,
} from "../schemas/profile-chat"
import {
  AuthenticationRequiredError,
  AuthService,
  InvalidTokenError,
} from "../services/auth-service"
import {
  ProfileAgentConfigurationError,
  ProfileAgentRequestError,
} from "../services/profile-agent-client"
import { ProfileChatService } from "../services/profile-chat-service"

export function createProfileChatRoutes(
  authService: AuthService,
  profileChatService: ProfileChatService
) {
  return new Elysia({
    name: "profile-chat-routes",
    prefix: "/profile-chat",
  }).post(
    "/message",
    async ({ body, headers, status }) => {
      try {
        const currentUser = await authService.authenticateAuthorizationHeader(
          headers.authorization
        )

        return await profileChatService.chat(currentUser, body)
      } catch (error) {
        if (
          error instanceof AuthenticationRequiredError ||
          error instanceof InvalidTokenError
        ) {
          return status(401, {
            detail: "Invalid or expired token",
          })
        }

        if (error instanceof ProfileAgentConfigurationError) {
          return status(500, {
            detail: error.message,
          })
        }

        if (error instanceof ProfileAgentRequestError) {
          return status(502, {
            detail: error.message,
          })
        }

        console.error("Profil chat endpoint'inde beklenmeyen hata", error)

        return status(500, {
          detail:
            "Beklenmeyen bir sunucu hatası oluştu. Lütfen tekrar deneyin.",
        })
      }
    },
    {
      body: ProfileChatRequestSchema,
      response: {
        200: ProfileChatResponseSchema,
        401: ErrorResponseSchema,
        502: ErrorResponseSchema,
        500: ErrorResponseSchema,
      },
    }
  )
}

File: apps/server/app/api/routes/profile_chat.py
# Module: Exposes the authenticated profile collection chat endpoint.
from fastapi import APIRouter, HTTPException, status

from app.api.deps import CurrentUserDep, ProfileChatServiceDep
from app.schemas.profile_chat import ProfileChatRequest, ProfileChatResponse
from app.services.profile_agent import (
    MissingLLMConfigurationError,
    ProfileAgentError,
    ProfileAgentTimeoutError,
)

router = APIRouter(prefix="/profile-chat")


@router.post("/message", response_model=ProfileChatResponse)
def send_profile_chat_message(
    request: ProfileChatRequest,
    current_user: CurrentUserDep,
    profile_chat_service: ProfileChatServiceDep,
) -> ProfileChatResponse:
    try:
        return profile_chat_service.chat(current_user, request)
    except MissingLLMConfigurationError as error:
        raise HTTPException(
            status_code=status.HTTP_503_SERVICE_UNAVAILABLE,
            detail=str(error),
        ) from error
    except ProfileAgentTimeoutError:
        raise HTTPException(
            status_code=status.HTTP_502_BAD_GATEWAY,
            detail=(
                "Profil asistanı zamanında yanıt veremedi. "
                "Lütfen kısa bir süre sonra tekrar deneyin."
            ),
        ) from None
    except ProfileAgentError:
        raise HTTPException(
            status_code=status.HTTP_502_BAD_GATEWAY,
            detail="Profil asistanı yanıt üretemedi. Lütfen tekrar deneyin.",
        ) from None
